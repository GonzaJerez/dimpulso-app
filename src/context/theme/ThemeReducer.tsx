import { darkTheme, lightTheme } from "../../constants/Colors";
import { ThemeActions, ThemeState } from "../../interfaces/theme-interfaces";



export const ThemeReducer = (state:ThemeState, actions: ThemeActions):ThemeState => {
    switch (actions.type) {
        case "set_dark_theme":
            return {...darkTheme}

        case "set_light_theme":
            return {...lightTheme}
    
        default:
            return state;
    }
}