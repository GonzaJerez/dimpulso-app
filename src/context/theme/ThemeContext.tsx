import { createContext, useEffect, useReducer } from "react";
import { useColorScheme } from "react-native";
import AsyncStorage from '@react-native-async-storage/async-storage'

import { darkTheme, lightTheme } from "../../constants/Colors";
import { ThemeState } from "../../interfaces/theme-interfaces";
import { ThemeReducer } from "./ThemeReducer";

interface ThemeContextProps {
    theme: ThemeState,
    setDarkTheme: () => void,
    setLightTheme: () => void,
}


export const ThemeContext = createContext({} as ThemeContextProps);

export const ThemeProvider = ({children}:{children:JSX.Element | JSX.Element[]}) => {

    // retorna el tema por defecto
    const colorScheme = useColorScheme();

    const [theme, dispatch] = useReducer(ThemeReducer, (colorScheme === 'dark') ? darkTheme : lightTheme);

    /**
   * Recupera el tema elegido anteriormente por el usuario
   */
  useEffect(() => {
    getThemeFromAsyncStorage()
      .then(resp => {
        (resp === 'light')
          ? setLightTheme()
          : setDarkTheme()
      })
  }, [])

  const getThemeFromAsyncStorage = async () => {
    const themeStorage = await AsyncStorage.getItem('theme') || 'light'
    return themeStorage
  }


  const setDarkTheme = async () => {
    await AsyncStorage.setItem('theme', 'dark')
    dispatch({type: 'set_dark_theme'})

  }

  const setLightTheme = async () => {
    await AsyncStorage.setItem('theme', 'light')
    dispatch({type: 'set_light_theme'})
  }
    return (
        <ThemeContext.Provider
            value={{
                theme,
                setDarkTheme,
                setLightTheme
            }}
        >
            {children}
        </ThemeContext.Provider>
    )
}