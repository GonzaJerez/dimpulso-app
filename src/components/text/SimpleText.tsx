import { Text } from 'react-native'
import React, { useContext } from 'react'
import { ThemeContext } from '../../context/theme/ThemeContext'
import { TextComponentProps } from '../../interfaces/textsComponents-interface';

export const SimpleText = ({children, style}:TextComponentProps) => {

  const {theme} = useContext(ThemeContext);

  return (
      <Text style={{color: theme.colors.text, ...style}}>{children}</Text>
  )
}
