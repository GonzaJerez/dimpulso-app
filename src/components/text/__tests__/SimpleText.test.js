import renderer from "react-test-renderer";
import { SimpleText } from "../SimpleText";

import { ThemeContext } from "../../../context/theme/ThemeContext";

describe("<SimpleText />", () => {
  it("should be render successfully", () => {
    const component = renderer
      .create(
        <ThemeContext.Provider value={{ theme: { colors: { text: "black" } } }}>
          <SimpleText>Hola</SimpleText>
        </ThemeContext.Provider>
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  it("should be render successfully with custom styles", () => {
    const component = renderer
      .create(
        <ThemeContext.Provider value={{ theme: { colors: { text: "black" } } }}>
          <SimpleText style={{ marginTop: 50 }}>Hola</SimpleText>
        </ThemeContext.Provider>
      )
      .toJSON();

    expect(component.props.style.marginTop).toBe(50);
  });
});
