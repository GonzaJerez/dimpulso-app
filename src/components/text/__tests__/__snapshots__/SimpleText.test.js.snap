// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SimpleText /> should be render successfully 1`] = `
<Text
  style={
    Object {
      "color": "black",
    }
  }
>
  Hola
</Text>
`;
